alexandria/core/views.py

```diff
@@ -168,87 +171,63 @@ def multi(self, request, **kwargs):
 
             return response
 
-    @action(methods=["post"], detail=True, url_path="generate-thumbnail")
-    def generate_thumbnail(self, request, pk=None):
-        if not settings.ALEXANDRIA_ENABLE_THUMBNAIL_GENERATION:
-            raise ValidationError(_("Thumbnail generation is not enabled."))
-
-        file = self.get_object()
-        if file.variant == models.File.THUMBNAIL:
-            raise ValidationError(
-                _("File is already a thumbnail, cannot generate thumbnail.")
-            )
-        elif file.renderings.all().exists():
-            raise ValidationError(
-                _("File already has thumbnail, cannot generate multiple thumbnails.")
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        if settings.ALEXANDRIA_ENABLE_AT_REST_ENCRYPTION:
+            serializer.validated_data[
+                "encryption_status"
+            ] = settings.ALEXANDRIA_ENCRYPTION_METHOD
+        obj = serializer.save()
+        if (
+            settings.ALEXANDRIA_ENABLE_THUMBNAIL_GENERATION
+            and obj.variant == models.File.Variant.ORIGINAL
+        ):
+            try:
+                create_thumbnail(obj)
+            except DjangoCoreValidationError as e:
+                raise ValidationError(*e.messages)
```

Comment:
Here's a semantic difference between the new version and the old: The old would upload a file and then just be unable to create a thumbnail. Here, we'll reject the whole upload if the thumbnail fails for some reason.

Should we instead just log an error if thumbnail creation fails? Cause otherwise, we'll sure have some files that just can't be uploaded, with potentially nothing the user can do about it

Reply:
I would say that raising the ValidationError at this point `L189` the code will not roll back the previous `serializer.save()` statement, no? Therefore the upload will still be successful for the actual object but no thumbnail will be available. I agree that it might be confusing for the frontend, since there is no success message for the upload but only an error message regarding the thumbnail.

So, no sure what's the right thing to do here.. The main object has been created. But only a thorough success of creation and configured side effects like thumbnail creation would suggest a return value of 2xx. So maybe I could add to the message that the primary object was created but the thumbnail remained somewhere in a pickle..

Reply:
That depends on whether [ATOMIC_REQUESTS](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#tying-transactions-to-http-requests) is enabled or not. In any case I think that when we return an error, the front-end should reasonably expect that the whole request did not go through.

If we store the main file, we should return a proper response, independent of the thumbnail. If we decide that thumbnail creation is integral part of the upload, we should wrap the whole thing in a transaction and rollback if thumbnail creation fails. Followup thought: If we save the thumbnail successfully, it should be in the relationships in the output, and if it failed, there won't be a thumbnail relationship.

Generally I think we should not make the thumbnail a critical element. There may be documents that can't be thumbnailed despite being valid. If we're concerned about the main document being messed up, we should do an explicit validation for that and not rely on the thumbnail success

Reply:
Fair points. So I'll do just that: if thumbnail creation was successful, include it in the returned object date. I'll just log an error on failing thumbnail creation if that's configured, so they may be picked up if occurring frequently.

Reply:
Followup: the thumbnail is no direct relation of the original.. so that won't work. I will however return a 201 if the original was created successfully. The other option would be to implement a one shot document with files upload endpoint where you could add a new (version of a) file to the document list.. but that would require some more specs.